add API.md (#1)
amperka/GpioExpanderPi: API.md

# GPIO Expander Pi API

## `class GpioExpander`

Create an object `GpioExpander` to communicate with a particular board based on Amperka GPIO expander. The end-user libraries often wrap (or inherit) this class to provide domain-specific high level hardware interaction methods.

### `GpioExpander(i2c_address=42)`

Constructor. The `i2c_address` specifies the board I²C address which is `42` factory-default but can be changed programmatically.

### `pin_mode(pin: int, mode: PinMode) -> None`

Configures the operation `mode` of the `pin`:

- PinMode.INPUT
- PinMode.INPUT_PULLUP
- PinMode.INPUT_PULLDOWN
- PinMode.OUTPUT

### `analog_read(pin: int) -> float`

Reads the analog value from one pin of board. Return value in range from `0.0` to `1.0`.

### `digital_read(pin: int) -> bool`

> Reads the bit value from one pin of board. Return `True` or `False` value.

Reviewer: ```suggestion
Reads a digital signal value from a pin of the board. Returns `True` or `False`.
```

> Reads the bit value from one pin of board. Return `GPIO.HIGH` or `GPIO.LOW` value. This values defined as `True` and `False` respectively.

Reviewer: ```suggestion
Reads the bit value from one pin of the board. Return `GPIO.HIGH` or `GPIO.LOW` value. These values defined as `True` and `False` respectively.
```

So… why `GPIO.HIGH` and `GPIO.LOW` are ever required? Why can’t they be just `True` or `False`. OK, Arduino API does use HIGH and LOW, but this is _not_ Arduino and the de-facto standard `gpiozero` library does not use HIGH/LOW, etc.

Suggest KISS: just use `bool`.

Reviewer 2: Done

### `digital_write(pin: int, value: bool) -> None`

> Write `True` or `False` value to a `pin`.

Reviewer: ```suggestion
Outputs digital signal (`True` or `False`) on a `pin`.
```

> Write `GPIO.HIGH` or `GPIO.LOW` value to a `pin`. This values defined as `True` and `False` respectively.

Reviewer: Ditto

Reviewer 2: Done

### `analog_write(pin: int, value: float) -> None`

Outputs PMW signal on a `pin`. The `value` argument specifies the duty cycle and should be in the range from `0.0` to `1.0`

### `change_address(new_address: int) -> None`

Changes the I²C address of the module. The change is in effect only while the board is powered on. If you want to save it permanently call the `save_address` method.

### `save_address() -> None`

Permanently saves the current board I²C address.

### `set_pwm_freq(freq: int) -> None`

> Sets the PWM frequency for the analog output. The `freq` must be in range from `30`(Hz) to `65535`(Hz).

Reviewer: ```suggestion
Sets a PWM frequency (in hertz) for all PWM outputs of the GPIO expander. The `freq` argument should be in the range from `30` to `65535`.
```

Reviewer 2: Done

# GPIO Expander Pi API

## `class GpioExpander`

Create an object `GpioExpander` to communicate with a particular board based on Amperka GPIO expander. The end-user libraries often wrap (or inherit) this class to provide domain-specific high level hardware interaction methods.

### `GpioExpander(i2c_address=42)`

Constructor. The `i2c_address` specifies the board I²C address which is `42` factory-default but can be changed programmatically.

### `pin_mode(pin: int, mode: int) -> None`

Configures the operation `mode` of the `pin`:

- GPIO.INPUT
- GPIO.INPUT_PULLUP
- GPIO.INPUT_PULLDOWN

> - GPIO.OUTPUT

Reviewer: Why `GPIO`? Looks too general. One of possible pythonic ways:

```py
from enum import Enum

class PinMode(Enum):
    INPUT = 1   #  use enum.auto() if the end-value “1” does not matter much
    INPUT_PULLUP = 2
    INPUT_PULLDOWN = 123
    OUTPUT = 3
```

Reviewer 2: Done

These constants are static members of the `GPIO` class.

### `analog_read(pin: int) -> float`

Reads the analog value from one pin of board. Return value in range from `0` to `1.0`.

### `digital_read(pin: int) -> bool`

Reads the bit value from one pin of board. Return `GPIO.HIGH` or `GPIO.LOW` value. This values defined as `True` and `False` respectively.

### `digital_write(pin: int, value: bool) -> None`

Write `GPIO.HIGH` or `GPIO.LOW` value to a `pin`. This values defined as `True` and `False` respectively.

### `analog_write(pin: int, value: float) -> None`

> Writes an analog value from `0` to `1.0` (PWM wave) to a `pin`.

Reviewer: ```suggestion
Outputs PMW signal on a `pin`. The `value` argument specifies the duty cycle and should be in the range from `0.0` to `1.0`
```

### `change_address(new_address: int) -> None`

Changes the I²C address of the module. The change is in effect only while the board is powered on. If you want to save it permanently call the `save_address` method.

### `save_address() -> None`

Permanently saves the current board I²C address.

> ### `pwm_freq(value: int) -> None`

Reviewer: ```suggestion
### `set_pwm_freq(value: int) -> None`
```

> Sets the PWM frequency for the analog output.

Reviewer: Information about valid (and sane) values of `value` is missing. Can I specify 12345678 here to output the signal with exactly 12345678 herz frequency? I doubt so.

Reviewer 2: Done

# GPIO Expander Pi API

## `class GpioExpander`

> Create an object `GpioExpander` to communicate with a particular GpioExpander boards.

Reviewer: ```suggestion
Create an object `GpioExpander` to communicate with a particular board based on Amperka GPIO expander. The end-user libraries often wrap (or inherit) this class to provide domain-specific high level hardware interaction methods.
```

### `GpioExpander(i2c_address=42)`

> The constructor for GpioExpander class. The `i2c_address` specifies the board

Reviewer: ```suggestion
Constructor. The `i2c_address` specifies the board
```

I²C address which is `42` factory-default but can be changed programmatically.

> ### `pin_mode(pin: int, mode: int) -> None`

Reviewer: `pin: int` ← what if I want to address `A3`? What `int` should I pass here?

Reviewer 2: All boards based on the GPIO Expander have pins without alphabetic numbers.

Reviewer: Ah, OK

> ### `pin_mode(pin: int, value: int) -> None`

Reviewer 3: Change` "value" to "mode" in the parameter list.
```suggestion
### `pin_mode(pin: int, mode: int) -> None`
```

> Configures state the `pin`:

Reviewer 3: ```suggestion
Configures the operation `mode` of the `pin`:
```

- INPUT
- INPUT_PULLUP
- INPUT_PULLDOWN

> - OUTPUT

Reviewer: What’s this? True enum, semi-enum, globals? Where are they defined? Use fully qualified names, please.

Reviewer 2: Can we create a separate class `GPIO` where these constants will be static members and put it in a separate module?

Reviewer 3: I think we can add these constants as static members directly to the GpioExpander class. Then we can call the `pin_mode` method as:
`obj.pin_mode(pin, GpioExpander.INPUT)`

### `analog_read(pin: int) -> float`

Reads the analog value from one pin of board. Return value in range from `0` to `1.0`.

### `digital_read(pin: int) -> int`

Reads the bit value from one pin of board. Return `0` or `1` value.

> ### `digital_write(pin: int) -> None`

Reviewer 4: ### `digital_write(pin: int, value: int) -> None`

Reviewer 2: I didn’t notice, thanks.

> Write `1` or `0` value to a `pin`.

Reviewer: Should use booleans, not integers

Reviewer 2: We suggest adding to the GPIO class, which will contain all the constants, two static members: `HIGH` and `LOW` equal `True` and `False`, respectively. Then `value` in `digital_write` and the return value `digital_read` is changed to bool.

Reviewer: Why so complicated? Just `True` and `False` is fine

### `analog_write(pin: int, value: int) -> None`

> Writes an analog value from `0` to `255` (PWM wave) to a `pin`.

Reviewer: Inconsistency: should be “from 0.0 to 1.0”, not the “0..255” cargo cult

Reviewer 2: Done
